### PR TITLE
[Feature] Add swin-transformer model.

### PR DESCRIPTION
**Pull request by mzr1996**
Reproduce swin-transformer model and training recipes.

### CODE REVIEW

File: mmcls/models/backbones/swin_transformer.py [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+        B, L, C = x.shape
+        assert L == H * W, 'input feature has wrong size'
+        assert H % 2 == 0 and W % 2 == 0, f'x size ({H}*{W}) are not even.'
+
+        x = x.view(B, H, W, C)
+
+        x0 = x[:, 0::2, 0::2, :]  # B H/2 W/2 C
+        x1 = x[:, 1::2, 0::2, :]  # B H/2 W/2 C
+        x2 = x[:, 0::2, 1::2, :]  # B H/2 W/2 C
+        x3 = x[:, 1::2, 1::2, :]  # B H/2 W/2 C
```
Comment (xvjiarui) [inline mmcls/models/backbones/swin_transformer.py:50]: Maybe consider `nn.Unfold`?

File: mmcls/models/backbones/swin_transformer.py [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+        x = self.sampler(x)  # B, 4*C, H/2*W/2
+        x = x.transpose(1, 2)  # B, H/2*W/2, 4*C
+
+        x = self.norm(x) if self.norm else x
+        x = self.reduction(x)
+
+        return x
+
+
+class PatchEmbed(BaseModule):
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:57]: Duplicated in ViT

File: mmcls/models/backbones/swin_transformer.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+
+        self.register_buffer('attn_mask', attn_mask)
+
+    def forward(self, query, key=None, value=None, residual=None, **kwargs):
+        if key is None:
+            key = query
+        if value is None:
+            value = key
+        if residual is None:
+            residual = query
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:308]: `residual` has changed to `identity` in MMCV.

File: mmcls/models/backbones/swin_transformer.py [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+                                              float(-100.0)).masked_fill(
+                                                  attn_mask == 0, float(0.0))
+        else:
+            attn_mask = None
+
+        self.register_buffer('attn_mask', attn_mask)
+
+    def forward(self, query, key=None, value=None, residual=None, **kwargs):
+        if key is None:
+            key = query
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:304]: Not used.

File: mmcls/models/backbones/swin_transformer.py [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+        if value is None:
+            value = key
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:306]: Not used.

File: mmcls/models/backbones/swin_transformer.py [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+
+        if self.downsample:
+            query = self.downsample(query)
+        return query
+
+
+@BACKBONES.register_module()
+class SwinTransformer(BaseBackbone):
+    """ Swin Transformer
+    A PyTorch impl of : `Swin Transformer:
```
Comment (ZwwWayne) [inline mmcls/models/backbones/swin_transformer.py:474]: impl -> implementation, should reduce abbreviations in docstring.

File: configs/_base_/datasets/imagenet_bs128_swin_384.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# dataset settings
```
Comment (Junjun2016) [inline configs/_base_/datasets/imagenet_bs128_swin_384.py:1]:
```suggestion
# dataset settings
# Only for evaluation
```

File: mmcls/models/backbones/swin_transformer.py [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+from mmcv.runner.base_module import BaseModule, ModuleList
+
+from ..builder import BACKBONES
+from ..utils import PatchEmbed, PatchMerging, to_2tuple
+from .base_backbone import BaseBackbone
+
+
+@ATTENTION.register_module()
+class WindowMSA(BaseModule):
+    """Window based multi-head self attention (W-MSA) module with relative
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:19]:
```suggestion
    """Window based multi-head self-attention (W-MSA) module with relative
```

File: mmcls/models/backbones/swin_transformer.py [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+        # Handle auto padding
+        self.auto_pad = auto_pad
+        if self.auto_pad:
+            self.pad_r = (self.window_size -
+                          W % self.window_size) % self.window_size
+            self.pad_b = (self.window_size -
+                          H % self.window_size) % self.window_size
+            dummy = torch.empty((1, H, W, 1))  # 1 H W 1
+            dummy = F.pad(dummy, (0, 0, 0, self.pad_r, 0, self.pad_b))
+            _, self.H_pad, self.W_pad, _ = dummy.shape
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:171]: Why not use `self.H_pad = H + self.pad_b` instead?

File: mmcls/models/backbones/swin_transformer.py [hunk 10]
```diff
@@ -0,0 +1,544 @@
...
+        # reverse cyclic shift
+        if self.shift_size > 0:
+            x = torch.roll(
+                shifted_x,
+                shifts=(self.shift_size, self.shift_size),
+                dims=(1, 2))
+        else:
+            x = shifted_x
+
+        if self.auto_pad and self.pad_r or self.pad_b:
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:250]:
```suggestion
        if self.pad_r or self.pad_b:
```

File: mmcls/models/backbones/swin_transformer.py [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+    @staticmethod
+    def double_step_seq(step1, len1, step2, len2):
+        seq1 = torch.arange(0, step1 * len1, step1)
+        seq2 = torch.arange(0, step2 * len2, step2)
+        return (seq1[:, None] + seq2[None, :]).reshape(1, -1)
+
+
+@ATTENTION.register_module()
+class ShiftWindowMSA(BaseModule):
+
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:130]: It's better to add docstring.

File: mmcls/models/backbones/swin_transformer.py [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+        B, H, W, C = x.shape
+        window_size = self.window_size
+        x = x.view(B, H // window_size, window_size, W // window_size,
+                   window_size, C)
+        windows = x.permute(0, 1, 3, 2, 4, 5).contiguous()
+        windows = windows.view(-1, window_size, window_size, C)
+        return windows
+
+
+class SwinBlock(BaseModule):
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:276]: It's better to add docstring.

File: mmcls/models/backbones/swin_transformer.py [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+        x = self.attn(x)
+        x = x + identity
+
+        identity = x
+        x = self.norm2(x)
+        x = self.ffn(x, identity=identity)
+        return x
+
+
+class SwinBlockSequence(BaseModule):
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:332]: It's better to add docstring.

File: mmcls/models/backbones/swin_transformer.py [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+                 downsample=None,
+                 drop_path=0.,
+                 norm_cfg=dict(type='LN'),
+                 block_cfg=dict(),
+                 auto_pad=False,
+                 init_cfg=None):
+        super().__init__(init_cfg)
+
+        if not isinstance(drop_path, list):
+            drop_path = [deepcopy(drop_path) for _ in range(depth)]
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:348]:
```suggestion
            drop_path = [drop_path for _ in range(depth)]
```

File: mmcls/models/backbones/swin_transformer.py [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+
+class SwinBlockSequence(BaseModule):
+
+    def __init__(self,
+                 embed_dims,
+                 input_resolution,
+                 depth,
+                 num_heads,
+                 downsample=None,
+                 drop_path=0.,
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:340]:
```suggestion
                 drop_paths=0.,
```

File: mmcls/models/backbones/swin_transformer.py [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+        use_abs_pos_embed (bool): If True, add absolute position embedding to
+            the patch embedding. Defaults to False.
+        auto_pad (bool): If True, auto pad feature map to fit window_size.
+            Defaults to False.
+        norm_cfg (dict, optional): Config dict for normalization layer at end
+            of backone. Defaults to dict(type='LN')
+        stage_cfg (dict, optional): Extra config dict for stages.
+            Defaults to None.
+        patch_cfg (dict, optional): Extra config dict for patch embedding.
+            Defaults to None.
```
Comment (Junjun2016) [inline mmcls/models/backbones/swin_transformer.py:415]: It's better to give examples.

**Comment by ycxioooong** [comment]
Please resolve the conflicts, thanks.

**Review by ycxioooong — APPROVED** [review]
LGTM

**Pull request merged**